File: 032_Substring with Concatenation of All Words.py
"""
Substring with Concatenation of All Words

Given a string s and an array of words of equal length, return all starting indices
of substrings that are a concatenation of each word exactly once and without intervening chars.
"""

from typing import List

def find_substring(s: str, words: List[str]) -> List[int]:
    if not s or not words: return []
    word_len = len(words[0])
    word_count = len(words)
    total = word_len * word_count
    if len(s) < total: return []
    from collections import Counter
    need = Counter(words)
    res = []

    for i in range(word_len):
        left = i
        window = Counter()
        count = 0
        for j in range(i, len(s) - word_len + 1, word_len):
            w = s[j:j+word_len]
            if w in need:
                window[w] += 1
                count += 1
                while window[w] > need[w]:
                    left_w = s[left:left+word_len]
                    window[left_w] -= 1
                    left += word_len
                    count -= 1
                if count == word_count:
                    res.append(left)
            else:
                window.clear()
                count = 0
                left = j + word_len
    return res


if __name__ == '__main__':
    print('Concat 1:', find_substring('barfoothefoobarman', ['foo','bar']))
    print('Concat 2:', find_substring('wordgoodgoodgoodbestword', ['word','good','best','word']))

# Time Complexity: O(n * word_len)
# Space Complexity: O(number of distinct words)

File: 032_Substring with Concatenation of All Words.cpp
/*
Substring with Concatenation of All Words

Given a string s and an array of words of equal length, return all starting indices of substrings
that are a concatenation of each word exactly once and without any intervening characters.
*/

#include <bits/stdc++.h>
using namespace std;

vector<int> findSubstring(string s, vector<string>& words) {
    vector<int> res;
    if (words.empty() || s.empty()) return res;
    int wordLen = words[0].size(), wordCount = words.size(), total = wordLen * wordCount;
    if ((int)s.size() < total) return res;
    unordered_map<string,int> need;
    for (auto &w: words) need[w]++;

    for (int i = 0; i < wordLen; ++i) {
        int left = i, count = 0;
        unordered_map<string,int> window;
        for (int j = i; j + wordLen <= (int)s.size(); j += wordLen) {
            string w = s.substr(j, wordLen);
            if (need.count(w)) {
                window[w]++;
                ++count;
                while (window[w] > need[w]) {
                    string leftw = s.substr(left, wordLen);
                    window[leftw]--;
                    --count;
                    left += wordLen;
                }
                if (count == wordCount) res.push_back(left);
            } else {
                window.clear();
                count = 0;
                left = j + wordLen;
            }
        }
    }
    return res;
}

int main() {
    string s1 = "barfoothefoobarman";
    vector<string> w1 = {"foo","bar"};
    auto r1 = findSubstring(s1, w1);
    cout << "Concat 1:";
    for (int x: r1) cout << ' ' << x;
    cout << endl; // 0 9

    string s2 = "wordgoodgoodgoodbestword";
    vector<string> w2 = {"word","good","best","word"};
    auto r2 = findSubstring(s2, w2);
    cout << "Concat 2:";
    for (int x: r2) cout << ' ' << x;
    cout << endl; // empty

    return 0;
}

// Time Complexity: O(n * wordLen)
// Space Complexity: O(number of distinct words)
